colossalai/shardformer/layer/_operation.py

```diff
@@ -164,24 +176,162 @@ def backward(ctx, grad_output):
                 handle = dist.all_reduce(grad_input, group=ctx.process_group, async_op=True)
             # Relay on CUDA_DEVICE_MAX_CONNECTIONS=1 to have
             # all-reduce scheduled first and have GPU resources allocated, CUDA_DEVICE_MAX_CONNECTIONS=1 is set in shardformer.py
-
         if _grad_accum_fusion_available and weight.grad is not None:
             grad = weight.grad
-            if grad.dtype == torch.float32:
-                fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp32(total_input, grad_output, grad)
-                grad_weight = None
-            elif grad.dtype == torch.float16:
-                fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp16(total_input, grad_output, grad)
+            if use_zbv:
+                # TODO: append input, grad_output_, weight, grad func to WeightGradStore
+                if grad.dtype == torch.float32:
+                    WeightGradStore.put(
+                        total_input,
+                        grad_output,
+                        weight,
+                        functools.partial(
+                            execute_w_pass_grad_accum,
+                            wgrad_gemm_accum_func=fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp32,
+                        ),
+                    )
+                    grad_weight = None
+                elif grad.dtype in (torch.float16, torch.bfloat16):
+                    WeightGradStore.put(
+                        total_input,
+                        grad_output,
+                        weight,
+                        functools.partial(
+                            execute_w_pass_grad_accum,
+                            wgrad_gemm_accum_func=fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp16,
+                        ),
+                    )
```

Comment:
Might be more concise to do this?
```
wgrad_accum_func = fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp32 if grad.dtype == torch.float32 else fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp16
...
functools.partial(
execute_w_pass_grad_accum,
wgrad_gemm_accum_func = wgrad_accum_func
)
```

Reply:
> Might be more concise to do this?
> 
> ```
> wgrad_accum_func = fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp32 if grad.dtype == torch.float32 \
>     else fused_weight_gradient_mlp_cuda.wgrad_gemm_accum_fp16
> ...
> functools.partial(
> wgrad_gemm_accum_func = wgrad_accum_func
> )
> ```

Hesitating on whether or not to make changes, But good to head u man  : ))

Reply:
🦾